src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/CDSPlugin.java

```diff
@@ -0,0 +1,127 @@
+import jdk.tools.jlink.plugin.PluginException;
+import jdk.tools.jlink.plugin.ResourcePool;
+import jdk.tools.jlink.plugin.ResourcePoolBuilder;
+
+/**
+ *
+ * CDS plugin
+ */
+public final class CDSPlugin extends AbstractPlugin implements PostProcessor {
+    private static final String NAME = "generate-cds-archive";
+    private Platform targetPlatform;
+    private Platform runtimePlatform;
+
+    public CDSPlugin() {
+        super(NAME);
+    }
+
+
+    private String javaExecutableName() {
+        if (targetPlatform.os() == Platform.OperatingSystem.WINDOWS) {
+            return "java.exe";
+        } else {
+            return "java";
+        }
+    }
+
+    private void generateCDSArchive(ExecutableImage image, boolean noCoops) {
+        List<String> javaCmd = new ArrayList<String>();
+        Path javaPath = image.getHome().resolve("bin").resolve(javaExecutableName());
+        if (!Files.exists(javaPath)) {
+            throw new PluginException("Cannot find java executable at: " + javaPath.toString());
+        }
+        javaCmd.add(javaPath.toString());
+        javaCmd.add("-Xshare:dump");
+        String archiveMsg = "CDS";
+        if (noCoops) {
+            javaCmd.add("-XX:-UseCompressedOops");
+            archiveMsg += "-NOCOOPS";
+        }
+        ProcessBuilder builder = new ProcessBuilder(javaCmd);
+        int status = -1;
+        try {
+            Process p = builder.inheritIO().start();
+            status = p.waitFor();
+        } catch (Exception ex) {
+            ex.printStackTrace();
+        }
+        if (status == 0) {
+            System.out.println("Created " + archiveMsg + " archive successfully");
+        } else {
+            throw new PluginException("Failed creating " + archiveMsg + " archive!");
+        }
+    }
+
+    @Override
+    public List<String> process(ExecutableImage image) {
+        targetPlatform = image.getTargetPlatform();
+        runtimePlatform = Platform.runtime();
+
+        if (!targetPlatform.equals(runtimePlatform)) {
+            throw new PluginException("Cannot generate CDS archives: target image platform " +
+                    targetPlatform.toString() + " is different from runtime platform " +
+                    runtimePlatform.toString());
+        }
+
+        Path classListPath = image.getHome().resolve("lib").resolve("classlist");
+        if (Files.exists(classListPath)) {
+            generateCDSArchive(image,false);
+
+            if (image.is64Bit()) {
+                generateCDSArchive(image,true);
+            }
+        } else {
+            throw new PluginException("Cannot generate CDS archives: classlist not found: " +
+                                      classListPath.toString());
+        }
+        return null;
+    }
+
+    @Override
+    public Category getType() {
+        return Category.PROCESSOR;
+    }
+
+    @Override
+    public ResourcePool transform(ResourcePool in, ResourcePoolBuilder out) {
+        in.transformAndCopy((file) -> {
```

Comment:
This method should not be called in a post-processor.    Plugin API needs some re-thinking to support post-processor plugin.   As `Plugin::transform` is abstract method, for now this method should simply throw `UnsupportedOperationException`.

Reply:
It is being called. The minimal implementation is:

```
    @Override
    public ResourcePool transform(ResourcePool in, ResourcePoolBuilder out) {
        return in;
    }
```
I've filed [JDK-8272734](https://bugs.openjdk.java.net/browse/JDK-8272743) to follow-up the issue.

Reply:
Thanks for filing the issue.   I didn't expect `Plugin::transform` should be called.  This is a good workaround for now.